---
title: "Lee et al. 2020 example"
author: "Won-Min Song"
date: '2023-11-30'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## MSC applications to influenza, COVID-19 infected PBMC scRNA-seq

This is R codes used to identify unique cell subsets associated to the infections, as shown in the manuscript. The pre-processed data in Seurat object format can be downloaded from NF Data Portal and on Synapse with DOI, [https://doi.org/10.7303/syn52966803](https://doi.org/10.7303/syn52966803), under Lee_et_al_2020 folder. These data should be placed under "Data/infectious_diseases" folder.

```{r Set up,message=FALSE}
rm(list = ls())
library(Seurat)
library(DESeq2)
library(Matrix.utils)
library(Matrix)
library(doParallel)
library(igraph)
library(scran)
library(ggraph)
library(MSC)
library(ggplot2)
library(ggrepel)
library(patchwork)
library(cowplot)

out.dir <- "Lee_et_al";dir.create(out.dir)
#seu = readRDS("Data/infectious_diseases/integrated_data.seurat_wt_celltype.RDS")
seu = readRDS("/media/won-min/My Passport/SingleCell/InfectiousDisease/Flu_and_COVID19/Lee_et_al/integrated_data.seurat_wt_celltype.RDS")

# update infection group
vec = seu@meta.data$subject.group
ii = which(!is.na(seu@meta.data$subject.status))
vec[ii] = seu@meta.data$subject.status[ii]
seu@meta.data$subject.group.name = factor(vec)

# get color maps
disease.col = c("healthy control" = "chartreuse1","Influenza patient" = "darkorchid1",
                "Asymptomatic case of COVID-19 patient" = "pink",
                "mild COVID-19 patient" = "deeppink",
                "severe COVID-19 patient" = "red")



```

## Cell type inference vis SingleR
We will use SingleR, and jointly use Monaco and HPCA collection to infer the cell types

```{r cell type inference, message=F}

library(SingleR)
library(celldex)

hpca.se <-  MonacoImmuneData()
#hpca.se <- BlueprintEncodeData()

m = GetAssayData(seu,assay = "SCT",slot = "data")
rownames(m) = gsub(":(.*)$","",rownames(m))
gene.common = intersect(rownames(hpca.se),rownames(m))
hpca.se = hpca.se[gene.common,]
m = m[gene.common,]

# fine subtype
pred.ct <- SingleR(test = m, ref = hpca.se, assay.type.test=1,
                   labels = hpca.se$label.fine)
seu@meta.data$inferred.cell.type.fine = pred.ct$pruned.labels[match(colnames(seu),rownames(pred.ct))]

# major cell type
pred.ct <- SingleR(test = m, ref = hpca.se, assay.type.test=1,
                   labels = hpca.se$label.main)

seu@meta.data$inferred.cell.type.broad = pred.ct$pruned.labels[match(colnames(seu),rownames(pred.ct))]

png(file = paste0(out.dir,"/monaco_inferred_celltypes.png"),res = 500,width = 5000,height = 4000)
print(DimPlot(seu,group.by = "inferred.cell.type.broad",reduction = "umap") + theme(plot.title = element_blank()))
dev.off()

#### re analyze progenitor cells
seu.sub = subset(seu,inferred.cell.type.broad == "Progenitors" | inferred.cell.type.broad == "Basophils")
hpca.se = HumanPrimaryCellAtlasData()
hpca.se = hpca.se[,hpca.se$label.main %in% c("Erythroblast","Platelets","HSC_-G-CSF","HSC_CD34+","MEP","Myelocyte")]

m = GetAssayData(seu.sub,assay = "SCT",slot = "data")
rownames(m) = gsub(":(.*)$","",rownames(m))
gene.common = intersect(rownames(hpca.se),rownames(m))
hpca.se = hpca.se[gene.common,]
m = m[gene.common,]

# broad subtype
pred.ct <- SingleR(test = m, ref = hpca.se, assay.type.test=1,
                   labels = hpca.se$label.main)
seu@meta.data$inferred.cell.type.broad[match(rownames(pred.ct),colnames(seu))] = pred.ct$pruned.labels

pred.ct <- SingleR(test = m, ref = hpca.se, assay.type.test=1,
                   labels = hpca.se$label.fine)
table(pred.ct$pruned.labels)
seu@meta.data$inferred.cell.type.fine[match(rownames(pred.ct),colnames(seu))] = pred.ct$pruned.labels

seu@meta.data$inferred.cell.type.fine[seu@meta.data$inferred.cell.type.fine %in% c("Progenitor cells","Basophils")] = NA
```

## MSC workflow

Run MSC on the integrated data. As Lee et al 2020 data is large (> 50,000 cells), we will remove cells with high dropout rates by scImpute framework, instead of imputing. Currently, the entire clustering workflow takes about one day with 8 cores. The result files in .RDS format are available from the Synapse/NF Data portal repository (MSC results: MSC_results.RDS, Dropout rate inference results: dropout_rate_results.RDS), and can be placed under 'Lee_et_al' folder to speed up running the vignette. 

```{r MSC workflow, echo=FALSE}
# thresholds for significant genes to use for clustering
pval.cutoff = 0.05 # p-value cutoff from gene dispersion evaluation
min.cells = 10 # minimum number of cells expressing a gene
bio.cutoff = 0 # minimum dispersion value
dthresh = 0.5 # minimum proportion of dropout reads per gene, per cell by scImpute
dropprop = 0.1 # 

n.cores = 6; # number of cores for parallel computation in MSC workflow

outfile = paste0(out.dir,"/MSC_results.RDS")

# if output file is not detected, run MSC workflow. 
if (!file.exists(outfile))
{
  #### use integrated data 
  mat = GetAssayData(seu,assay = "integrated")
  cnt = GetAssayData(seu,assay = "SCT",slot = "counts");cnt = cnt[match(rownames(mat),rownames(cnt)),]
  
  #### handle errors by negative values, then calculate the dispersions
  ep = 0
  if (any(mat < 0)) ep = abs(min(mat,na.rm = TRUE))
  gene.var = modelGeneVar(mat+ep,block = factor(seu@meta.data$orig.ident))
  ncell = rowSums(cnt > 1E-320,na.rm = T)
  gene.var$cell.express = ncell;
  gene.var$bio.filtered = rownames(gene.var) %in% rownames(subset(gene.var,bio > bio.cutoff & p.value < pval.cutoff & cell.express >= min.cells))
  
  
  #### evaluate dropout
  cat("Check dropout rates per gene...\n")
  if (!file.exists(paste0(out.dir,"/dropout_rate_results.RDS")))
  {
    drop.rates = get_dropout_rates(mat = cnt,gene.names = rownames(cnt))
    saveRDS(drop.rates,file = paste0(out.dir,"/dropout_rate_results.RDS"))
  }else{
    drop.rates = readRDS(paste0(out.dir,"/dropout_rate_results.RDS"))
  }
  n.dropout = rowSums(drop.rates < dthresh,na.rm = TRUE)
  rate.dropout = n.dropout/ncol(drop.rates)
  	
  # update gene.var with dropout data
  gene.var$cell.wt.dropout = n.dropout[match(rownames(gene.var),names(n.dropout))]
  gene.var$dropout.rate = rate.dropout[match(rownames(gene.var),names(rate.dropout))]
  gene.var$dropout.filtered = gene.var$dropout.rate > dropprop
  sig.gene.var = subset(gene.var,bio.filtered & dropout.filtered)
  
  # finalize remaining genes
  dgenes = rownames(sig.gene.var)
  cat(paste0("- number of genes filtered:",length(dgenes),"\n"))
  		
  # create binary matrix by dropout screening
  drop.matrix = drop.rates[match(dgenes,rownames(drop.rates)),]
  ii = drop.matrix <= dthresh
  
  # refine matrices: counts and normalized
  cat("Refine gene list and calculate pairwise correlations...\n")
  cnt = cnt[match(dgenes,rownames(cnt)),]
  cnt[!ii] = 0
  
  # get binary matrix for jaccard calculation
  bcnt = cnt;
  bcnt[bcnt > 1E-320] = 1
  
  # get normalized data
  mat = mat[match(dgenes,rownames(mat)),]
  rm(drop.matrix,drop.rates)
  gc()
  
  ## Run the msc_workflow() wrapper function to execute MSC workflow. 
  iter.res = msc_workflow(dat = mat,bcnt = bcnt,n.cores = n.cores,min.cells = min.cells, valid.gate = "both")
  iter.res = prune_hierarchy.compact(iter.res)
  
  saveRDS(iter.res,file = outfile)
  rm(mat,bcnt)
  gc()

}else{
  iter.res = readRDS(outfile)
  iter.res = prune_hierarchy.compact(iter.res)
}

```

## UMAP for results
```{r UMAP plotting,message = F}

# coordinate by umap
if (TRUE)
{
  xy = Embeddings(seu,"umap")
}

## get major cell type plots

vec = seu@meta.data$inferred.cell.type.broad;names(vec) = colnames(seu);
vec = vec[rownames(xy)];
singleton = names(table(vec))[table(vec) <= 10]
vec[vec %in% singleton] = NA
pobj.ct.broad = plot_dimred_manual.raw(xy,cell.feat =vec,alpha.val = 0.6,cls.col = NULL,add.labels = TRUE) + geom_text_repel(size = 13) + 
  guides(colour = guide_legend(ncol = 5,title = "Cell Types",label.theme = element_text(size = 15),title.theme = element_text(size = 17),
                               override.aes = list(alpha = 1,size = 4)))

# extract legend
library(ggpubr)
leg = get_legend(pobj.ct.broad) 
leg = as_ggplot(leg)

pobj.ct.broad = pobj.ct.broad + guides(colour = "none") + 
  theme_classic() + 
  theme(axis.line = element_blank(),axis.title = element_blank(),axis.text = element_blank(),axis.ticks = element_blank()) +
  annotate(geom = "segment",x = quantile(xy[,1],0),xend = quantile(xy[,1],0.1),y = min(xy[,2]),yend = min(xy[,2]),colour = "black",arrow = grid::arrow(length = unit(0.015, "npc"))) + 
  annotate(geom = "segment",x = min(xy[,1]),xend = min(xy[,1]),y = min(xy[,2]),yend = quantile(xy[,2],0.1),colour = "black",arrow = grid::arrow(length = unit(0.015, "npc")) )

pobj.ct.broad = pobj.ct.broad/leg + plot_layout(ncol = 1,heights = c(6,1))

## get fine cell type
vec = seu@meta.data$inferred.cell.type.fine;names(vec) = colnames(seu)
vec = vec[rownames(xy)]
singleton = names(table(vec))[table(vec) <= 10]
vec[vec %in% singleton] = NA
pobj.ct.fine = plot_dimred_manual.raw(xy,cell.feat =vec,alpha.val = 0.3,cls.col = NULL,add.labels = TRUE,is.repel =TRUE) + geom_text_repel(size = 13) + 
  guides(colour = guide_legend(ncol = 4,title = "Cell Types",label.theme = element_text(size = 12),title.theme = element_text(size = 17),
                               override.aes = list(alpha = 1,size = 4)))

# extract legend
library(ggpubr)
library(patchwork)
library(cowplot)
leg = get_legend(pobj.ct.fine) 
leg = as_ggplot(leg)

pobj.ct.fine = pobj.ct.fine + 
  guides(colour = "none") + 
  theme_classic() + 
  theme(axis.line = element_blank(),axis.title = element_blank(),axis.text = element_blank(),axis.ticks = element_blank()) +
  annotate(geom = "segment",x = quantile(xy[,1],0),xend = quantile(xy[,1],0.1),y = min(xy[,2]),yend = min(xy[,2]),colour = "black",arrow = grid::arrow(length = unit(0.015, "npc"))) + 
  annotate(geom = "segment",x = min(xy[,1]),xend = min(xy[,1]),y = min(xy[,2]),yend = quantile(xy[,2],0.1),colour = "black",arrow = grid::arrow(length = unit(0.015, "npc")) )

pobj.ct.fine = pobj.ct.fine/leg + plot_layout(ncol = 1,heights = c(6,3))

# put plots together
umap.obj = plot_grid(pobj.ct.broad,pobj.ct.fine,labels = c("A","B"),ncol = 1,rel_heights = c(1,1.5),label_fontface = "bold",label_size = 17)
```

## MSC results summary
This generates the UMAP plots for the first split in MSC, and the hierarchical piecharts
```{r MSC summary,message = F}
#### Pie plots & 
module.size = 50;
htbl = subset(iter.res$pruned.table,module.size >= 30 & module.size <= ncol(seu)*0.8)[,c("cluster.name","parent.cluster")]
modules = iter.res$modules

# for cell type
if (T)
{
  vec.ct = seu@meta.data$inferred.cell.type.broad;names(vec.ct) = colnames(seu);
  colmap.ct = c("B cells" = "blueviolet","NK cells" = "deeppink1","CD8+ T cells" = "brown","CD4+ T cells" = "slateblue","T cells" = "cadetblue1",
                "Progenitors" = "chocolate1","Basophils" = "darkgoldenrod1","Monocytes" = "darkseagreen4","Neutrophils" = "chartreuse",
                "Dendritic cells" = "red","Erythroblast" = "burlywood1","Platelets" = "lightblue","HSC_-G-CSF" = "blue","HSC_CD34+" = "deepskyblue","MEP" = "rosybrown")
  
  
  pie.ct = plot_hierarchy_pie(htbl,modules,vec.ct,colmap = colmap.ct,edge_colour = "grey") + 
    guides(fill = guide_legend(title.position = "top",title.hjust = 0.5,ncol = 4,title = "Cell Types",title.theme = element_text(size = 17),label.theme = element_text(size = 15)))
  txt.dat = pie.ct$data[,c("x","y","name")]
  txt.dat = subset(txt.dat,name %in% c(subset(iter.res$pruned.table,parent.cluster == "M1")$cluster.name))
  
  # extract legend
  library(ggpubr)
  leg = get_legend(pie.ct)
  leg = as_ggplot(leg)
  
  # add text and clean plot
  library(ggrepel)
  pie.ct = pie.ct + geom_text_repel(data = txt.dat,aes(x = x,y = y,label = name),colour = "darkblue",segment.colour = "darkblue",vjust = 1.5,size = 6) + 
    guides(fill = "none") + 
    theme(panel.border = element_blank(),legend.position = "right",legend.direction = "vertical",
          legend.text = element_text(size = 16),
          plot.title = element_text(size = 20,hjust = 0.5),
          plot.margin = margin(t = 0,r = 0,b = 0,l = 0)) 
  
  pie.ct = plot_grid(pie.ct,leg,ncol = 1,rel_heights = c(4,1))
}

# sample composition
if (T)
{
  vec.sc = seu@meta.data$subject.group.name;
  vec.sc = gsub(" patient| control|case of ","",vec.sc)
  names(vec.sc) = colnames(seu);
  colmap.sc = c("Influenza" = "darkorange","Asymptomatic COVID-19" = "brown","mild COVID-19" = "deeppink1","severe COVID-19" = "red",
                "healthy" = "chartreuse")
  
  
  pie.sc = plot_hierarchy_pie(htbl,modules,vec.sc,colmap = colmap.sc,edge_colour = "grey") + 
    guides(fill = guide_legend(title.position = "top",title.hjust = 0.5,nrow = 2,title = "Disease Status",title.theme = element_text(size = 17),label.theme = element_text(size = 15)))
  txt.dat = pie.sc$data[,c("x","y","name")]
  txt.dat = subset(txt.dat,name %in% c(subset(iter.res$pruned.table,parent.cluster == "M1")$cluster.name))
  
  # extract legend
  library(ggpubr)
  leg = get_legend(pie.sc)
  leg = as_ggplot(leg)
  
  # add text and clean plot
  library(ggrepel)
  pie.sc = pie.sc + geom_text_repel(data = txt.dat,aes(x = x,y = y,label = name),colour = "darkblue",segment.colour = "darkblue",vjust = 1.5,size = 6) + 
    guides(fill = "none") + 
    theme(panel.border = element_blank(),legend.position = "right",legend.direction = "vertical",
          legend.text = element_text(size = 16),
          plot.title = element_text(size = 20,hjust = 0.5),
          plot.margin = margin(t = 0,r = 0,b = 0,l = 0)) 
  
  pie.sc = plot_grid(pie.sc,leg,ncol = 1,rel_heights = c(5,1))
}

pie.collect = plot_grid(pie.ct,pie.sc,ncol = 1,labels = c("C","D"),label_size = 40)

# individual sample presence
if (T)
{
  vec.sa = seu@meta.data$sample;
  names(vec.sa) = colnames(seu);
  
  
  pie.sa = plot_hierarchy_pie(htbl,modules,vec.sa,colmap = NULL,edge_colour = "grey") 
  txt.dat = pie.sa$data[,c("x","y","name")]
  txt.dat = subset(txt.dat,name %in% c(subset(iter.res$pruned.table,parent.cluster == "M1")$cluster.name))
  
  # add text and clean plot
  library(ggrepel)
  pie.sa = pie.sa + geom_text_repel(data = txt.dat,aes(x = x,y = y,label = name),colour = "darkblue",segment.colour = "darkblue",vjust = 1.5,size = 6) + 
    guides(fill = "none") + 
    theme(panel.border = element_blank(),legend.position = "right",legend.direction = "vertical",
          legend.text = element_text(size = 16),
          plot.title = element_text(size = 20,hjust = 0.5),
          plot.margin = margin(t = 0,r = 0,b = 0,l = 0)) 
  png(file = paste0(out.dir,"/sample_composition.png"),res = 500,width = 5000,height = 5000)
  print(pie.sa)
  dev.off()
}

# add UMAP for first layer
if (TRUE)
{
  ### get visualization
  xy = Embeddings(seu,"umap")
  vec = get_rooted_membership(mtbl = iter.res$module.table,
                              modules = iter.res$modules,
                              pid = "M1",bg = colnames(seu))
  pobj.msc = plot_dimred_manual.raw(xy,
                                    cell.feat =vec,
                                    label.size = 11,
                                    alpha.val = 0.3,cls.col = NULL,add.labels = TRUE) +
    geom_text_repel(size = 13) + 
    guides(colour = guide_legend(ncol = 5,title = "Cell Types",label.theme = element_text(size = 15),title.theme = element_text(size = 17),
                                 override.aes = list(alpha = 1,size = 4)))
  
  pobj.msc = pobj.msc + guides(colour = "none") + 
    theme_classic() + 
    theme(axis.line = element_blank(),axis.title = element_blank(),axis.text = element_blank(),axis.ticks = element_blank()) +
    annotate(geom = "segment",x = quantile(xy[,1],0),xend = quantile(xy[,1],0.1),y = min(xy[,2]),yend = min(xy[,2]),colour = "black",arrow = grid::arrow(length = unit(0.015, "npc"))) + 
    annotate(geom = "segment",x = min(xy[,1]),xend = min(xy[,1]),y = min(xy[,2]),yend = quantile(xy[,2],0.1),colour = "black",arrow = grid::arrow(length = unit(0.015, "npc")) )
  
  rm(vec)
}

# add UMAP for major cell types
if (TRUE)
{
  ## get major cell type plots
  vec.ct = seu@meta.data$inferred.cell.type.broad;names(vec.ct) = colnames(seu);
  colmap.ct = c("B cells" = "blueviolet","NK cells" = "deeppink1","CD8+ T cells" = "brown","CD4+ T cells" = "slateblue","T cells" = "cadetblue1",
                "Progenitors" = "chocolate1","Basophils" = "darkgoldenrod1","Monocytes" = "darkseagreen4","Neutrophils" = "chartreuse",
                "Dendritic cells" = "red","Erythroblast" = "burlywood1","Platelets" = "lightblue","HSC_-G-CSF" = "blue","HSC_CD34+" = "deepskyblue","MEP" = "rosybrown")
  
  singleton = names(table(vec.ct))[table(vec.ct) <= 10]
  vec.ct[vec.ct %in% singleton] = NA
  pobj.ct.broad = plot_dimred_manual.raw(xy,cell.feat =vec.ct,alpha.val = 0.6,cls.col = NULL,label.size = 7,add.labels = TRUE) + geom_text_repel(size = 13) + 
    scale_colour_manual(values = colmap.ct) + 
    guides(colour = guide_legend(ncol = 5,title = "Cell Types",label.theme = element_text(size = 15),title.theme = element_text(size = 17),
                                 override.aes = list(alpha = 1,size = 4)))
  
  pobj.ct.broad = pobj.ct.broad + guides(colour = "none") + 
    theme_classic() + 
    theme(axis.line = element_blank(),axis.title = element_blank(),axis.text = element_blank(),axis.ticks = element_blank()) +
    annotate(geom = "segment",x = quantile(xy[,1],0),xend = quantile(xy[,1],0.1),y = min(xy[,2]),yend = min(xy[,2]),colour = "black",arrow = grid::arrow(length = unit(0.015, "npc"))) + 
    annotate(geom = "segment",x = min(xy[,1]),xend = min(xy[,1]),y = min(xy[,2]),yend = quantile(xy[,2],0.1),colour = "black",arrow = grid::arrow(length = unit(0.015, "npc")) )
  
}

umap.collect = plot_grid(pobj.msc,pobj.ct.broad,ncol = 1,labels = c("A","B"),label_size = 40)
upper.collect = plot_grid(umap.collect,pie.collect,ncol = 2,rel_widths = c(1.5,1,0.5),labels = c("",""),label_size = 40)

print(upper.collect)
```
